src/arithmetic/big_native/primes.rs

```diff
@@ -0,0 +1,518 @@
+
+    'nextrandom: for i in 0..reps {
+        let x = if i == reps - 1 && force2 {
+            BIG_2.clone()
+        } else {
+            gen_biguint_below(&mut rng, &nm3) + &*BIG_2
+        };
+
+        let mut y = x.modpow(&q, n);
+        if y.is_one() || y == nm1 {
+            continue;
+        }
+
+        for _ in 1..k {
+            y = y.modpow(&*BIG_2, n);
+            if y == nm1 {
+                break 'nextrandom;
+            }
+            if y.is_one() {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    true
+}
+
+/// Reports whether n passes the "almost extra strong" Lucas probable prime test,
+/// using Baillie-OEIS parameter selection. This corresponds to "AESLPSP" on Jacobsen's tables (link below).
+/// The combination of this test and a Miller-Rabin/Fermat test with base 2 gives a Baillie-PSW test.
+///
+///
+/// References:
+///
+/// Baillie and Wagstaff, "Lucas Pseudoprimes", Mathematics of Computation 35(152),
+/// October 1980, pp. 1391-1417, especially page 1401.
+/// http://www.ams.org/journals/mcom/1980-35-152/S0025-5718-1980-0583518-6/S0025-5718-1980-0583518-6.pdf
+///
+/// Grantham, "Frobenius Pseudoprimes", Mathematics of Computation 70(234),
+/// March 2000, pp. 873-891.
+/// http://www.ams.org/journals/mcom/2001-70-234/S0025-5718-00-01197-2/S0025-5718-00-01197-2.pdf
+///
+/// Baillie, "Extra strong Lucas pseudoprimes", OEIS A217719, https://oeis.org/A217719.
+///
+/// Jacobsen, "Pseudoprime Statistics, Tables, and Data", http://ntheory.org/pseudoprimes.html.
+///
+/// Nicely, "The Baillie-PSW Primality Test", http://www.trnicely.net/misc/bpsw.html.
+/// (Note that Nicely's definition of the "extra strong" test gives the wrong Jacobi condition,
+/// as pointed out by Jacobsen.)
+///
+/// Crandall and Pomerance, Prime Numbers: A Computational Perspective, 2nd ed.
+/// Springer, 2005.
+pub fn probably_prime_lucas(n: &BigUint) -> bool {
+    // println!("lucas: {}", n);
+    // Discard 0, 1.
+    if n.is_zero() || n.is_one() {
+        return false;
+    }
+
+    // Two is the only even prime.
+    if n.to_u64() == Some(2) {
+        return false;
+    }
+
+    // Baillie-OEIS "method C" for choosing D, P, Q,
+    // as in https://oeis.org/A217719/a217719.txt:
+    // try increasing P ≥ 3 such that D = P² - 4 (so Q = 1)
+    // until Jacobi(D, n) = -1.
+    // The search is expected to succeed for non-square n after just a few trials.
+    // After more than expected failures, check whether n is square
+    // (which would cause Jacobi(D, n) = 1 for all D not dividing n).
+    let mut p = 3u64;
+    let n_int = BigInt::from_biguint(Sign::Plus, n.clone());
+
+    loop {
+        if p > 10000 {
+            // This is widely believed to be impossible.
+            // If we get a report, we'll want the exact number n.
+            panic!("internal error: cannot find (D/n) = -1 for {:?}", n)
+        }
+
+        let j = jacobi(&BigInt::from(p * p - 4), &n_int);
```

Comment:
Jacobi is actually useful primitives that would be nice to expose (we use it in Class library) 
just saying it for future reference - not relevant to this PR